### PR TITLE
disable healthcheck for unsupported versions fixes #2359

### LINKED ISSUE #2359: Bad packet 254
### Describe the problem

Trying to host a Better than Adventure server, log keeps getting spammed. Doesn't happen outside on host (outside Docker). I'm assuming it's the included healthcheck script, which has always worked flawlessly for any release version I've tried. Is there perhaps an undocumented ENV variable for disabling it or something?

### Container definition

```yml
version: "3.8"

services:
  mc:
    image: itzg/minecraft-server:java8-multiarch
    tty: true
    stdin_open: true
    ports:
      - "25565:25565"
    environment:
      EULA: "true"
      VERSION: "1.0"

      MEMORY: "6G"
      USE_AIKAR_FLAGS: "true"

      TYPE: "CUSTOM"
      CUSTOM_SERVER: "https://github.com/Better-than-Adventure/bta-download-repo/releases/download/v1.7.7.0_01/better-than-adventure-1.7.7.0_01-server.jar"
    restart: unless-stopped
    volumes:
      # attach the relative directory 'data' to the container's /data path
      - ./data:/data
```

### Container logs

```
bta-docker-mc-1  | java.io.IOException: Bad packet id 254
bta-docker-mc-1  | 	at net.minecraft.core.net.packet.Packet.readPacket(Packet.java:86)
bta-docker-mc-1  | 	at net.minecraft.core.net.NetworkManager.readPacket(NetworkManager.java:136)
bta-docker-mc-1  | 	at net.minecraft.core.net.NetworkManager.readNetworkPacket(NetworkManager.java:257)
bta-docker-mc-1  | 	at net.minecraft.core.net.NetworkReaderThread.run(NetworkReaderThread.java:31)
bta-docker-mc-1  | 2023-08-24 17:55:15 INFO  Minecraft - /127.0.0.1:54506 lost connection
bta-docker-mc-1  | java.io.IOException: Bad packet id 254
bta-docker-mc-1  | 	at net.minecraft.core.net.packet.Packet.readPacket(Packet.java:86)
bta-docker-mc-1  | 	at net.minecraft.core.net.NetworkManager.readPacket(NetworkManager.java:136)
bta-docker-mc-1  | 	at net.minecraft.core.net.NetworkManager.readNetworkPacket(NetworkManager.java:257)
bta-docker-mc-1  | 	at net.minecraft.core.net.NetworkReaderThread.run(NetworkReaderThread.java:31)
bta-docker-mc-1  | 2023-08-24 17:55:20 INFO  Minecraft - /127.0.0.1:59354 lost connection
bta-docker-mc-1  | java.io.IOException: Bad packet id 254
bta-docker-mc-1  | 	at net.minecraft.core.net.packet.Packet.readPacket(Packet.java:86)
bta-docker-mc-1  | 	at net.minecraft.core.net.NetworkManager.readPacket(NetworkManager.java:136)
bta-docker-mc-1  | 	at net.minecraft.core.net.NetworkManager.readNetworkPacket(NetworkManager.java:257)
bta-docker-mc-1  | 	at net.minecraft.core.net.NetworkReaderThread.run(NetworkReaderThread.java:31)
bta-docker-mc-1  | 2023-08-24 17:55:25 INFO  Minecraft - /127.0.0.1:59364 lost connection
bta-docker-mc-1  | java.io.IOException: Bad packet id 254
bta-docker-mc-1  | 	at net.minecraft.core.net.packet.Packet.readPacket(Packet.java:86)
bta-docker-mc-1  | 	at net.minecraft.core.net.NetworkManager.readPacket(NetworkManager.java:136)
bta-docker-mc-1  | 	at net.minecraft.core.net.NetworkManager.readNetworkPacket(NetworkManager.java:257)
bta-docker-mc-1  | 	at net.minecraft.core.net.NetworkReaderThread.run(NetworkReaderThread.java:31)
bta-docker-mc-1  | 2023-08-24 17:55:30 INFO  Minecraft - /127.0.0.1:54022 lost connection
bta-docker-mc-1  | java.io.IOException: Bad packet id 254
bta-docker-mc-1  | 	at net.minecraft.core.net.packet.Packet.readPacket(Packet.java:86)
bta-docker-mc-1  | 	at net.minecraft.core.net.NetworkManager.readPacket(NetworkManager.java:136)
bta-docker-mc-1  | 	at net.minecraft.core.net.NetworkManager.readNetworkPacket(NetworkManager.java:257)
bta-docker-mc-1  | 	at net.minecraft.core.net.NetworkReaderThread.run(NetworkReaderThread.java:31)
bta-docker-mc-1  | 2023-08-24 17:55:36 INFO  Minecraft - /127.0.0.1:37490 lost connection
bta-docker-mc-1  | java.io.IOException: Bad packet id 254
bta-docker-mc-1  | 	at net.minecraft.core.net.packet.Packet.readPacket(Packet.java:86)
bta-docker-mc-1  | 	at net.minecraft.core.net.NetworkManager.readPacket(NetworkManager.java:136)
bta-docker-mc-1  | 	at net.minecraft.core.net.NetworkManager.readNetworkPacket(NetworkManager.java:257)
bta-docker-mc-1  | 	at net.minecraft.core.net.NetworkReaderThread.run(NetworkReaderThread.java:31)
bta-docker-mc-1  | 2023-08-24 17:55:41 INFO  Minecraft - /127.0.0.1:37498 lost connection
bta-docker-mc-1  | java.io.IOException: Bad packet id 254
bta-docker-mc-1  | 	at net.minecraft.core.net.packet.Packet.readPacket(Packet.java:86)
bta-docker-mc-1  | 	at net.minecraft.core.net.NetworkManager.readPacket(NetworkManager.java:136)
bta-docker-mc-1  | 	at net.minecraft.core.net.NetworkManager.readNetworkPacket(NetworkManager.java:257)
bta-docker-mc-1  | 	at net.minecraft.core.net.NetworkReaderThread.run(NetworkReaderThread.java:31)
bta-docker-mc-1  | 2023-08-24 17:55:46 INFO  Minecraft - /127.0.0.1:58850 lost connection
bta-docker-mc-1  | java.io.IOException: Bad packet id 254
bta-docker-mc-1  | 	at net.minecraft.core.net.packet.Packet.readPacket(Packet.java:86)
bta-docker-mc-1  | 	at net.minecraft.core.net.NetworkManager.readPacket(NetworkManager.java:136)
bta-docker-mc-1  | 	at net.minecraft.core.net.NetworkManager.readNetworkPacket(NetworkManager.java:257)
bta-docker-mc-1  | 	at net.minecraft.core.net.NetworkReaderThread.run(NetworkReaderThread.java:31)
bta-docker-mc-1  | 2023-08-24 17:55:51 INFO  Minecraft - /127.0.0.1:58862 lost connection
bta-docker-mc-1  | java.io.IOException: Bad packet id 254
bta-docker-mc-1  | 	at net.minecraft.core.net.packet.Packet.readPacket(Packet.java:86)
bta-docker-mc-1  | 	at net.minecraft.core.net.NetworkManager.readPacket(NetworkManager.java:136)
bta-docker-mc-1  | 	at net.minecraft.core.net.NetworkManager.readNetworkPacket(NetworkManager.java:257)
bta-docker-mc-1  | 	at net.minecraft.core.net.NetworkReaderThread.run(NetworkReaderThread.java:31)
bta-docker-mc-1  | 2023-08-24 17:55:56 INFO  Minecraft - /127.0.0.1:38022 lost connection
bta-docker-mc-1  | java.io.IOException: Bad packet id 254
bta-docker-mc-1  | 	at net.minecraft.core.net.packet.Packet.readPacket(Packet.java:86)
bta-docker-mc-1  | 	at net.minecraft.core.net.NetworkManager.readPacket(NetworkManager.java:136)
bta-docker-mc-1  | 	at net.minecraft.core.net.NetworkManager.readNetworkPacket(NetworkManager.java:257)
bta-docker-mc-1  | 	at net.minecraft.core.net.NetworkReaderThread.run(NetworkReaderThread.java:31)
bta-docker-mc-1  | 2023-08-24 17:56:01 INFO  Minecraft - /127.0.0.1:38038 lost connection
bta-docker-mc-1  | java.io.IOException: Bad packet id 254
bta-docker-mc-1  | 	at net.minecraft.core.net.packet.Packet.readPacket(Packet.java:86)
bta-docker-mc-1  | 	at net.minecraft.core.net.NetworkManager.readPacket(NetworkManager.java:136)
bta-docker-mc-1  | 	at net.minecraft.core.net.NetworkManager.readNetworkPacket(NetworkManager.java:257)
bta-docker-mc-1  | 	at net.minecraft.core.net.NetworkReaderThread.run(NetworkReaderThread.java:31)
bta-docker-mc-1  | 2023-08-24 17:56:06 INFO  Minecraft - /127.0.0.1:40746 lost connection
```

### PR DESCRIPTION
**Pull request by lightlike**
fixes #2359

### CODE REVIEW

File: scripts/start-finalExec [hunk 1]
```diff
@@ -221,7 +221,11 @@ function copyFilesForCurseForge() {
   cp -f /data/eula.txt "${FTB_DIR}/"
 }
 
-if versionLessThan 1.7; then
+if versionLessThan 'b1.8'; then
```
Comment (itzg) [inline scripts/start-finalExec:224]: Now I'm confused by Mojang's version scheme. I thought beta was  before release 1.0. So I was expecting the 1.7 comparison to remain and you would add a new comparisons for < 1.3 which implies beta releases. However, the version comparison logic doesn't know about "b" prefix...unless you're enhancing that separately.

**Comment by itzg** [comment]
image

File: scripts/start-finalExec (hunk 1 above)
Comment (lightlike) [reply scripts/start-finalExec:224]: I thought your version comparison uses [launchermeta](https://launchermeta.mojang.com/mc/game/version_manifest_v2.json) which has the Versions including Alpha and Beta in there and is/can be ordered.

I really did not look that far into that script. If you want to, I can but that will take some time.

**Comment by lightlike** [comment]
I do not know what versioning you were expecting but I am pretty sure Minecraft uses the [Semantic Versioning Standard](https://semver.org/).
Which is in itself very simple to use. (Creating those versions is another point. That is horrible to explain to another developer that does not really know the system in detail)

Minecraft just used the prefixes `a` and `b` to seperate those versions. It is as if the major version (first number) would have been 0 for Alpha, 1 for Beta and 2 for Release (that would have made more sense from a versioning perspective but would confuse consumers)

**Comment by itzg** [comment]
Your comment makes it sound like I had control over Mojang's version scheme since you say "you" several times. I am quite aware of semantic versioning and **they**, Mojang, (not me) only started using proper semantic versioning at 1.0. Their alpha and beta version numbers violate https://semver.org/#spec-item-9

I'll queue this up to work on later, but FYI the version comparison uses Maven's version parser in https://github.com/itzg/mc-image-helper/blob/master/src/main/java/me/itzg/helpers/versions/CompareVersionsCommand.java

**Comment by itzg** [comment]
> This means the the version b1.7.3 is older than the Release version 1.0.

...and yes, that was the point I was originally trying to convey.

**Comment by lightlike** [comment]
Sorry. I did not want to imply that you were the one making those versions.
That was my regional german coming though. We sometimes say "you" with the meaning of "someone" or similar.

I did find that version comparison but completely oversaw that you were using some maven versioning object to compare these versions.
I thought you were just loading the json and comparing them via the position in the list or by date.
If you are doing a real parse and compare then this really will not work.

I will try looking if there is a good way to implement a way for this to compare with alpha and beta.
I'll probably go in and do a check if the version string starts with `a`, `b` or a number and only call the comparison if they are both the same.

**Comment by lightlike** [comment]
Should this work for the comaprison or do you want another implementation?
https://github.com/lightlike/mc-image-helper/blob/master/src/main/java/me/itzg/helpers/versions/CompareVersionsCommand.java#L23

**Comment by itzg** [comment]
Sorry for overreacting 😬

Yes, I was roughly thinking the same where some preprocessing of the version string could be done prior to the maven/semvar comparisons.

I'm making some other changes in mc-image-helper right now, so I'll take a stab at the changes there and then you can update this PR to use that.

**Comment by lightlike** [comment]
I already made a PR for the changes in the comparison.
I tried to keep it readable and in line with your other code.

You can use that if you want to.

**Comment by itzg** [comment]
Bump the version to [1.34.7](https://github.com/itzg/mc-image-helper/releases/tag/1.34.7) here and you can finalize the integration:

https://github.com/itzg/docker-minecraft-server/blob/03034d2f152004e8028e37a77850eb2d8294b994/Dockerfile#L45

File: scripts/start-finalExec (hunk 1 above)
Comment (itzg) [inline scripts/start-finalExec:224]: Re-reading the protocol docs again, I think the existing condition for < 1.7 is still correct given https://wiki.vg/Server_List_Ping#1.6, so please put that back as an `elif`

...and before that is where you could add a new condition for < 1.6 which will disable the health check...for now.

Also with re-reading the docs, I realized one could also improve the implementation in https://github.com/itzg/mc-monitor/blob/master/slp/slp.go to handle the 1.4-1.5 and the <=1.3 variations of the protocol messages.

**Review by itzg — CHANGES_REQUESTED** [review]
Thanks for your patience on this whole PR

File: scripts/start-finalExec (hunk 1 above)
Comment (lightlike) [reply scripts/start-finalExec:224]: The < 1.7 is still in there in line 228.

The check for 1.6 still works with older versions from b1.8 - 1.5 as the sent data is only additive.
I just tested it again with 1.4.7 and the existing healthcheck works without any Problem.
You can just use the 1.6 healthcheck for that.

Comment (itzg) [reply scripts/start-finalExec:224]: 🤦‍♂️sorry, I missed the `elif` on line 228.

**Review by itzg — APPROVED** [review]
Awesome!

**Comment by itzg** [comment]
I'll get this merged after the tests run.

**Comment by lightlike** [comment]
> Thanks for your patience on this whole PR

No problem.
That was actually pretty fast and with good reasons for it be stay open.
I currently have a PR open for over a year with another project.

**Pull request merged**